src/shogun/optimization/SGDMinimizer.cpp

```diff
@@ -0,0 +1,80 @@
+ /*
+ * Copyright (c) The Shogun Machine Learning Toolbox
+ * Written (w) 2015 Wu Lin
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ *    list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ * The views and conclusions contained in the software and documentation are those
+ * of the authors and should not be interpreted as representing official policies,
+ * either expressed or implied, of the Shogun Development Team.
+ *
+ */
+#include <shogun/optimization/SGDMinimizer.h>
+#include <shogun/optimization/GradientDescendUpdater.h>
+using namespace shogun;
+
+CSGDMinimizer::CSGDMinimizer()
+	:CFirstOrderStochasticMinimizer()
+{
+	init();
+}
+
+CSGDMinimizer::~CSGDMinimizer()
+{
+}
+
+CSGDMinimizer::CSGDMinimizer(CFirstOrderStochasticCostFunction *fun)
+	:CFirstOrderStochasticMinimizer(fun)
+{
+	init();
+}
+
+float64_t CSGDMinimizer::minimization()
+{
+	CMap<TParameter*, SGVector<float64_t> > *variable_reference=m_fun->obtain_variable_reference();
+	SG_REF(variable_reference);
+	CFirstOrderStochasticCostFunction *fun=dynamic_cast<CFirstOrderStochasticCostFunction *>(m_fun);
+	REQUIRE(fun,"the cost function must be a stochastic cost function\n");
+	for(;m_cur_passes<m_num_passes;m_cur_passes++)
+	{
+		fun->begin_sample();
+		while(fun->next_sample())
```

Comment:
this step can be paralleled (either synchronous or asynchronous)
for now it just a simple implementation.